src/records.rs

```diff
@@ -0,0 +1,303 @@
+    pub fn hold_until(mut self, until: u64) -> Self {
+        self.record.hold_until = Some(until);
+        self
+    }
+
+    /// Update the record to no longer be on hold
+    pub fn unhold(mut self) -> Self {
+        self.record.hold_until = None;
+        self
+    }
+
+    /// Save the record to the store.
+    ///
+    /// Returns the stored record.
+    pub fn save<L: Ledger>(&mut self) -> Result<Record, KeystoreError<L>> {
+        self.store.store(self.record.uid, &self.record)?;
+        Ok(self.record.clone())
+    }
+}
+
+impl<'a> Deref for RecordEditor<'a> {
+    type Target = Record;
+
+    fn deref(&self) -> &Record {
+        &self.record
+    }
+}
+
+impl<'a> DerefMut for RecordEditor<'a> {
+    fn deref_mut(&mut self) -> &mut Record {
+        &mut self.record
+    }
+}
+
+type RecordsStore = KeyValueStore<u64, Record>;
+
+pub struct Records {
+    store: RecordsStore,
+    /// Record (size, uid) indexed by asset type, owner, and freeze status, for easy allocation as
+    /// transfer or freeze inputs. The records for each asset are ordered by increasing size, which
+    /// makes it easy to implement a worst-fit allocator that minimizes fragmentation.
+    asset_records:
+        PersistableHashMapBTreeMultiMap<(AssetCode, UserAddress, FreezeFlag), (RecordAmount, u64)>,
+    /// Record uids indexed by nullifier, for easy removal when confirmed as transfer inputs
+    nullifier_records: PersistableHashMap<Nullifier, u64>,
+}
+
+impl Records {
+    #![allow(dead_code)]
+
+    pub fn new<L: Ledger>(
+        loader: &mut AtomicStoreLoader,
+        adaptor: EncryptingResourceAdapter<(u64, Option<Record>)>,
+        fill_size: u64,
+    ) -> Result<Self, KeystoreError<L>> {
+        let log = AppendLog::load(loader, adaptor, "keystore_records", fill_size)?;
+        let store = RecordsStore::new(log)?;
+        let mut records = Self {
+            store,
+            asset_records: Persistable::new(),
+            nullifier_records: Persistable::new(),
+        };
+        records.reload();
+        Ok(records)
+    }
+
+    /// Create a Record
+    ///
+    /// Returns an editor to the newly created Record
+    pub fn create<L: Ledger>(
+        &mut self,
+        uid: u64,
+        ro: RecordOpening,
+        nullifier: Nullifier,
+        hold_until: Option<u64>,
+    ) -> Result<RecordEditor<'_>, KeystoreError<L>> {
+        let record = Record {
+            ro,
+            uid,
+            nullifier,
+            hold_until,
+        };
+        let ro = &record.ro;
+        self.asset_records.insert((
+            (ro.asset_def.code, ro.pub_key.address(), ro.freeze_flag),
+            (record.ro.amount.into(), record.uid),
+        ));
+        self.nullifier_records
+            .insert((record.nullifier, record.uid));
+        let mut editor = RecordEditor::new(self, record);
+        editor.save()?;
+        Ok(editor)
+    }
+
+    /// Rebuild the indices for this store
+    fn reload(&mut self) {
```

Comment:
This function is supposed to act kind of like a revert (i.e. set everything in memory back to the state it has on disk) and this is how it is used in `new`. But, since it uses the `Persistable` interface to build the indexes, it ends up with `pending_changes` populated in both indices. This means if revert is called after loading and before the first commit, it will not only revert outstanding changes but will also clear out the indices.

I think `Transactions` has the same issue, as well.

Reply:
I will remove reload here and address the issue with `Transactions` in a different pr